### PR TITLE
Autofixing new Perf cops in 0.37.2

### PR DESCRIPTION
**Pull request by lamont-granquist**
6 Performance/Casecmp
  18 Performance/Detect
   1 Performance/RangeInclude
  27 Performance/RedundantBlockCall
   6 Performance/RedundantMatch
   5 Performance/RedundantMerge
  18 Performance/StringReplacement

### CODE REVIEW

**Comment by lamont-granquist** [comment]
travis is angrysauce about unrelated issues with a bad github-changelog-generator release

@chef/client-core thoughts?

File: lib/chef/provider/service/upstart.rb [hunk 1]
```diff
@@ -62,7 +62,7 @@ def initialize(new_resource, run_context)
           end
 
           platform, version = Chef::Platform.find_platform_and_version(run_context.node)
-          if platform == "ubuntu" && (8.04..9.04).include?(version.to_f)
+          if platform == "ubuntu" && (8.04..9.04).cover?(version.to_f)
```
Comment (danielsdeleo) [inline lib/chef/provider/service/upstart.rb]: What is this?

File: Gemfile [hunk 2]
```diff
@@ -30,7 +30,7 @@ group(:development, :test) do
   # gem 'chefstyle', github: 'chef/chefstyle'
   gem "chefstyle", git: "https://github.com/chef/chefstyle.git", branch: "master"
 
-  gem "ruby-shadow", platforms: :ruby unless RUBY_PLATFORM.downcase.match(/(aix|cygwin)/)
+  gem "ruby-shadow", platforms: :ruby unless RUBY_PLATFORM.downcase =~ /(aix|cygwin)/
```
Comment (danielsdeleo) [inline Gemfile]: Is there more explanation of this? I'm assuming it only kicks in if you don't use the match data return... Wondering how much of an actual improvement this is.

File: lib/chef/chef_fs/data_handler/container_data_handler.rb [hunk 3]
```diff
@@ -15,10 +15,10 @@ def preserve_key?(key)
           return key == "containername"
         end
 
-        def verify_integrity(object, entry, &on_error)
+        def verify_integrity(object, entry)
```
Comment (danielsdeleo) [inline lib/chef/chef_fs/data_handler/container_data_handler.rb]: I'm a bit hesitant about these ones. I personally like the `yield` keyword for non-logical reasons, but having the block argument makes the method signature more descriptive, especially for longer methods where the `yield` may be harder to spot on a quick reading. You can of course fix that with yard/rdoc/whatever but I don't know if we find that worthwhile.

Comment (lamont-granquist) [inline lib/chef/chef_fs/data_handler/container_data_handler.rb]: The block argument (alone) is 4x slower than just `yield`ing with no proc:

https://github.com/JuanitoFatas/fast-ruby#proccall-and-block-arguments-vs-yieldcode

Sorta feel like documentation should be fixed with YARD.

File: Gemfile (hunk 2 above)
Comment (lamont-granquist) [inline Gemfile]: Not a lot:  https://github.com/JuanitoFatas/fast-ruby#regexp-vs-stringmatch-vs-string-code-

I kind of like =~ better than match -- i've probably been infected by writing too many lines of Perl in my life though....

File: lib/chef/provider/service/upstart.rb (hunk 1 above)
Comment (lamont-granquist) [inline lib/chef/provider/service/upstart.rb]: IDK, good eye...

Comment (lamont-granquist) [inline lib/chef/provider/service/upstart.rb]: wow, this one is a lot faster:

https://github.com/JuanitoFatas/fast-ruby#range

although for Float, i'd think `#include?` would have to be an alias to `#cover?`

File: lib/chef/chef_fs/data_handler/container_data_handler.rb (hunk 3 above)
Comment (thommay) [inline lib/chef/chef_fs/data_handler/container_data_handler.rb]: I tend to agree we should do this with documentation. @lamont-granquist could you add some minimal YARDage to this PR for the things that used to take blocks?

File: Gemfile (hunk 2 above)
Comment (thommay) [inline Gemfile]: Personally, I think =~ is a bit more intention-revealing, but it's a bit of a wash.

**Comment by thommay** [comment]
I'm :+1: modulo documentation for things that are block-y

**Comment by danielsdeleo** [comment]
I'm okay with these then. TBH I don't think the block thing will actually change much for us unless we're calling these things in a tight enough loop to trigger significantly more GCs but as long as we add doc to make it clear that these methods require block arguments, it's not really a big deal.

:+1:

**Pull request merged**